File: src/experimental/testAdvectionPrototype.cpp
#include "experimental/DescriptorPrototype.hpp"
#include "AMReX.H"
#include "AMReX_ParallelDescriptor.H"
#include "AMReX_Print.H"
#include <cmath>

using quokka::experimental::AdvectionSimulationPrototype;
using quokka::experimental::HydroModule;
using quokka::experimental::PhysicsDescriptor;
using quokka::experimental::PhysicsLayout;
using quokka::experimental::ProblemHooks;
using quokka::experimental::setRuntimePhysicsLayout;

namespace
{

AMREX_FORCE_INLINE auto sawtooth(amrex::Real x, amrex::Real prob_lo, amrex::Real prob_hi) -> amrex::Real
{
	const amrex::Real length = prob_hi - prob_lo;
	amrex::Real shifted = x - prob_lo;
	shifted = std::fmod(shifted + length, length);
	return std::fmod(shifted + 0.5 * length, length);
}

} // namespace

auto main(int argc, char **argv) -> int
{
	amrex::Initialize(argc, argv);
	int status = 0;
	{
		constexpr amrex::Real prob_lo = 0.0;
		constexpr amrex::Real prob_hi = 1.0;
		constexpr int nx = 6400;
		constexpr amrex::Real advection_velocity = 1.0;
		constexpr amrex::Real max_time = 1.0;
		constexpr amrex::Real cfl = 0.4;
		constexpr amrex::Real err_tol = 9.0e-3;

		PhysicsLayout layout;
		layout.total_cc_components = 1;
		layout.rad_group_stride = 0;
		layout.scalar_stride = 1;
		setRuntimePhysicsLayout(layout);

		AdvectionSimulationPrototype sim(layout);
		sim.configureGrid(nx, prob_lo, prob_hi);
		sim.setAdvectionVelocity(advection_velocity, 0.0, 0.0);
		sim.setCfl(cfl);
		sim.setMaxTime(max_time);
		sim.setMaxTimesteps(5000);

		sim.setInitialCondition([](amrex::Real x) { return sawtooth(x, prob_lo, prob_hi); });

		sim.setExactSolution([=](amrex::Real x, amrex::Real time) {
			const amrex::Real length = prob_hi - prob_lo;
			amrex::Real x0 = x - advection_velocity * time;
			while (x0 < prob_lo) {
				x0 += length;
			}
			while (x0 >= prob_hi) {
				x0 -= length;
			}
			return sawtooth(x0, prob_lo, prob_hi);
		});

		ProblemHooks hooks;
		hooks.diagnostics = [&sim]() {
			if (amrex::ParallelDescriptor::IOProcessor()) {
				amrex::Print() << "Prototype time = " << sim.currentTime() << ", L1(error) = " << sim.errorNorm() << "\n";
			}
		};
		sim.setHooks(hooks);

		sim.run();
		const amrex::Real err = sim.errorNorm();
		if (amrex::ParallelDescriptor::IOProcessor()) {
			amrex::Print() << "Prototype advection error norm = " << err << "\n";
		}
		status = (err < err_tol) ? 0 : 1;
	}
	amrex::Finalize();
	return status;
}

File: inputs/python/test_advection_prototype.py
import math
import pyquokka

prob_lo = 0.0
prob_hi = 1.0
nx = 4096
velocity = 1.0
max_time = 1.0
cfl = 0.4
err_tol = 0.007

def sawtooth(x: float, prob_lo: float, prob_hi: float) -> float:
    length = prob_hi - prob_lo
    shifted = math.fmod(x - prob_lo + length, length)
    return math.fmod(shifted + 0.5 * length, length)
  
def exact_func(grid: pyquokka.Grid, time: float) -> None:
    dx = grid.dx[0]
    ilo, ihi = grid.i_range
    length = prob_hi - prob_lo
    for i in range(ilo, ihi + 1):
        x = prob_lo + (float(i) + 0.5) * dx
        x0 = x - velocity * time
        while x0 < prob_lo:
            x0 += length
        while x0 >= prob_hi:
            x0 -= length
        grid.set_state(i, sawtooth(x0, prob_lo, prob_hi))

def init_func(grid: pyquokka.Grid) -> None:
    exact_func(grid, 0.0)
  
def diagnostics(time: float, err: float) -> None:
    print(f"Prototype time = {time}, L1(error) = {err}")

def main() -> int:
    pyquokka.initialize()
    sim = pyquokka.AdvectionSimulation()
    sim.configure_grid(nx, prob_lo, prob_hi)
    sim.set_advection_velocity(velocity, 0.0, 0.0)
    sim.set_cfl(cfl)
    sim.set_max_time(max_time)
    sim.set_max_timesteps(5000)
    sim.set_python_hooks(initial=init_func, exact=exact_func, diagnostics=diagnostics)
    sim.run()
    err = sim.error_norm()
    print(f"Prototype advection error norm = {err}")
    pyquokka.finalize()
    return 0 if err < err_tol else 1

if __name__ == "__main__":
    raise SystemExit(main())
